src/Microsoft.ML.Data/Transforms/CopyColumnsTransform.cs

```diff
@@ -163,4 +169,189 @@ protected override Delegate GetGetterCore(IChannel ch, IRow input, int iinfo, ou
+        {
+            Contracts.CheckValue(inputSchema, nameof(inputSchema));
+            Contracts.CheckValue(inputSchema.Columns, nameof(inputSchema.Columns));
+            var originDic = inputSchema.Columns.ToDictionary(x => x.Name);
+            var resultDic = inputSchema.Columns.ToDictionary(x => x.Name);
+            foreach ((string source, string name) pair in _columns)
+            {
+                if (originDic.ContainsKey(pair.source))
+                {
+                    var col = originDic[pair.source].CloneWithNewName(pair.name);
+                    resultDic[pair.name] = col;
+                }
+                else
+                {
+                    throw Contracts.ExceptUserArg(nameof(inputSchema), $"{pair.source} not found in {nameof(inputSchema)}");
+                }
+            }
+            var result = new SchemaShape(originDic.Values.ToArray());
+            return result;
+        }
+    }
+
+    public sealed class CopyColumnsTransformer : ITransformer, ICanSaveModel
+    {
+        private readonly (string Source, string Name)[] _columns;
+        private readonly IHost _host;
+
+        private class CopyColumnsRowMapper : IRowMapper
+        {
+            private readonly ISchema _schema;
+            private readonly HashSet<int> _originalColumnSources;
+            private (string source, string name)[] _columns;
+
+            public CopyColumnsRowMapper(ISchema schema, (string source, string name)[] columns)
+            {
+                _schema = schema;
+                _columns = columns;
+                _originalColumnSources = new HashSet<int>();
+                var sources = new HashSet<string>();
+                foreach (var source in columns.Select(x => x.source))
+                    sources.Add(source);
+                for (int i = 0; i < _schema.ColumnCount; i++)
+                    if (sources.Contains(_schema.GetColumnName(i)))
+                        _originalColumnSources.Add(i);
+
+            }
+
+            public Delegate[] CreateGetters(IRow input, Func<int, bool> activeOutput, out Action disposer)
+            {
+                var result = new Delegate[_columns.Length];
+                int i = 0;
+                foreach (var column in _columns)
+                {
+                    // validate activeOutput.
+                    input.Schema.TryGetColumnIndex(column.source, out int colIndex);
+                    var type = input.Schema.GetColumnType(colIndex);
+
+                    result[i++] = Utils.MarshalInvoke(MakeGetter<int>, type.RawType, input, colIndex);
+                }
+                disposer = null;
+                return result;
+            }
+
+            private Delegate MakeGetter<T>(IRow row, int src) => row.GetGetter<T>(src);
+
+            public Func<int, bool> GetDependencies(Func<int, bool> activeOutput) => (col) => (activeOutput(col) && _originalColumnSources.Contains(col));
```

Comment:
>GetDependencies [](start = 35, length = 15)

this is wrong: the argument to activeOutput is between 0 and `GreateGetters().Length`. The argument to the resulting Func is between 0 and `_schema.ColumnCount`

Reply:
I suspect you use `_colMap` as old-to-new mapping, but in reality you populate the other way round.

---
In reply to: [211792129](https://github.com/dotnet/machinelearning/pull/706#discussion_r211792129) [](ancestors = 211792129)

Reply:
I need both I guess

---
In reply to: [212480267](https://github.com/dotnet/machinelearning/pull/706#discussion_r212480267) [](ancestors = 212480267,211792129)

Reply:
Old to new is one-to-many

---
In reply to: [212482646](https://github.com/dotnet/machinelearning/pull/706#discussion_r212482646) [](ancestors = 212482646,212480267,211792129)